### ISSUE TITLE
Qwen vocabulary size mismatch when using c++ tokenizer

### ISSUE THREAD

**Issue by haowhsu-quic** [pytorch/executorch#11807]
### 🐛 Describe the bug

We are trying to load tokenizer.json for Qwen in C++ using tokenizers::HFTokenizer with following patch on `executorch/backends/qualcomm/scripts/build.sh`:
```diff
diff --git a/backends/qualcomm/scripts/build.sh b/backends/qualcomm/scripts/build.sh
index 8099ecb3d..ff615b0c2 100755
--- a/backends/qualcomm/scripts/build.sh
+++ b/backends/qualcomm/scripts/build.sh
@@ -85,6 +85,7 @@ if [ "$BUILD_AARCH64" = true ]; then
         -DEXECUTORCH_BUILD_EXTENSION_FLAT_TENSOR=ON \
         -DEXECUTORCH_BUILD_EXTENSION_TENSOR=ON \
         -DEXECUTORCH_ENABLE_EVENT_TRACER=ON \
+        -DSUPPORT_REGEX_LOOKAHEAD=ON \
         -DQNN_SDK_ROOT=$QNN_SDK_ROOT \
         -DCMAKE_TOOLCHAIN_FILE=$ANDROID_NDK_ROOT/build/cmake/android.toolchain.cmake \
         -DANDROID_ABI='arm64-v8a' \
@@ -104,6 +105,7 @@ if [ "$BUILD_AARCH64" = true ]; then
         -DANDROID_ABI='arm64-v8a' \
         -DANDROID_PLATFORM=android-30 \
         -DCMAKE_PREFIX_PATH=$CMAKE_PREFIX_PATH \
+        -DSUPPORT_REGEX_LOOKAHEAD=ON \
         -DEXECUTORCH_BUILD_KERNELS_QUANTIZED=ON \
         -DCMAKE_FIND_ROOT_PATH_MODE_PACKAGE=BOTH \
         -DPYTHON_EXECUTABLE=$PYTHON_EXECUTABLE \
@@ -130,6 +132,7 @@ if [ "$BUILD_X86_64" = true ]; then
         -DEXECUTORCH_BUILD_DEVTOOLS=ON \
         -DEXECUTORCH_BUILD_EXTENSION_MODULE=ON \
         -DEXECUTORCH_BUILD_EXTENSION_DATA_LOADER=ON \
+        -DSUPPORT_REGEX_LOOKAHEAD=ON \
         -DEXECUTORCH_BUILD_EXTENSION_FLAT_TENSOR=ON \
         -DEXECUTORCH_BUILD_KERNELS_QUANTIZED=ON \
         -DEXECUTORCH_BUILD_EXTENSION_TENSOR=ON \
@@ -155,6 +158,7 @@ if [ "$BUILD_X86_64" = true ]; then
    cmake $PRJ_ROOT/$EXAMPLE_ROOT \
        -DCMAKE_BUILD_TYPE=$BUILD_TYPE \
        -DCMAKE_PREFIX_PATH=$CMAKE_PREFIX_PATH \
+       -DSUPPORT_REGEX_LOOKAHEAD=ON \
        -DCMAKE_FIND_ROOT_PATH_MODE_PACKAGE=BOTH \
        -DPYTHON_EXECUTABLE=$PYTHON_EXECUTABLE \
        -B$EXAMPLE_ROOT
```

The number of vocabs is incorrect. Qwen tokenizer should have 151936 vocabs, but the tokenizer in c++ shows there are only 151665 vocabs from parsing the tokenizer.json.
We would like to know if there are any tips on loading Qwen tokenizer.json in C++.

We have tried the following 3 methods and non of them are working:
- Download tokenizer.json from hugging face
- Saving tokenizer.json from AutoTokenizer.from_pretrained("Qwen/Qwen2.5-0.5B").save_pretrained("tokenizer.json")
- Follow qwen3 README under Executorch and use the tokenizer.json under ~/.cache/huggingface/hub/models--Qwen--Qwen3-0.6B/snapshots/a9c98e602b9d36d2a2f7ba1eb0f5f31e4e8e5143/tokenizer.json

We would like to know will this discrepancy impact the accuracy. Thank you.

### Versions

Hi @cccclai, I create an issue here for following discussion. Thank you in advance for helping us.
Add @winskuo-quic for awareness.

**Comment by jackzhxng** [comment]
@haowhsu-quic can you comment the missing vocab entries?

**Comment by winskuo-quic** [comment]
Hi @jackzhxng,
I think there are currently 3 places that we check the number of vocabs.
1) From python's model config, which it shows there are 151936 tokens. Reproduce step:
   `config = Qwen2Config.from_pretrained("Qwen/Qwen2.5-0.5B")`
   `print(config.vocab_size)`
2) From python's tokenizer, which when I print the tokenizer, it will show all vocabs, and there's 151665 vocabs.
  `tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-0.5B")`
  `print(tokenizer)`
3) From c++'s tokenizers::HFTokenizer, we check the vocab size, which also shows there are 151665 vocabs.
  `tokenizer_->vocab_size()`

In python, I also tried to decode ids between 151665-151936, and it seems like these vocabs does not exist.
Based on the tests above, I think the c++ tokenizer is working as expected after turning on `-DSUPPORT_REGEX_LOOKAHEAD=ON`.
However, it seems like there's still a mismatch between model's config and the tokenizer. I would like to know if this is also happening on your side and which `vocab_size` should we be using when lowering the model with ExecuTorch.
Thanks

**Comment by jackzhxng** [comment]
Hi @winskuo-quic, a bit confused since you are mentioning both Qwen2.5 and Qwen3. Regardless, our tokenizer (and most likely the tokenizer returned by AutoTokenizer) is reporting the true vocab size, and both of these models should have similar values of around 151665. I'm not sure where 151936 is coming from, but if it's coming from an official source such as `Qwen2Config.from_pretrained("Qwen/Qwen2.5-0.5B")`, then it must representing the size of the vocabulary with padding tokens which the Qwen team added for distributed pretraining. For inference I wouldn't worry about this mismatch, unless you are seeing that our tokenizer's vocab size is greater compared to other official sources.

**Comment by winskuo-quic** [comment]
> Hi [@winskuo-quic](https://github.com/winskuo-quic), a bit confused since you are mentioning both Qwen2.5 and Qwen3. Regardless, our tokenizer (and most likely the tokenizer returned by AutoTokenizer) is reporting the true vocab size, and both of these models should have similar values of around 151665. I'm not sure where 151936 is coming from, but if it's coming from an official source such as `Qwen2Config.from_pretrained("Qwen/Qwen2.5-0.5B")`, then it must representing the size of the vocabulary with padding tokens which the Qwen team added for distributed pretraining. For inference I wouldn't worry about this mismatch, unless you are seeing that our tokenizer's vocab size is greater compared to other official sources.

Thanks for the response.
Qwen3 is mentioned at summary section of this issue because Executorch used qwen3 as an example to demonstrate where to get tokenizer.json. I am using the same way to retrieve tokenizer for qwen2.
The number 151936 actually comes from the official source.
If these extra vocabs are just for distributed pretraining, I think it shouldn't matter much.
Thanks again for the help on this issue.

**Comment by jackzhxng** [comment]
Glad to help!
